pallets/tendermint-light-client/src/lib.rs

```diff
@@ -0,0 +1,293 @@
+        /// ubiquitous event type
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+        /// Maximal number of finalized headers to keep in the storage, last-in first-out
+        #[pallet::constant]
+        type HeadersToKeep: Get<u32>;
+
+        /// time provider type, used to gauge whether blocks are within the trusting period
+        type TimeProvider: UnixTime;
+    }
+
+    #[pallet::pallet]
+    #[pallet::storage_version(STORAGE_VERSION)]
+    pub struct Pallet<T>(_);
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// Pallet is halted
+        LightClientHalted,
+        /// Pallet operations are resumed        
+        LightClientResumed,
+        /// Light client is initialized
+        LightClientInitialized,
+        /// New block has been verified and imported into storage \[relayer_address, imported_block_hash\]
+        ImportedLightBlock(T::AccountId, BridgedBlockHash),
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// Unable to deserialize extrinsic
+        DeserializeError,
+        /// Light client has not been initialized        
+        NotInitialized,
+        /// Light client has already been initialized
+        AlreadyInitialized,
+        /// Light client is currently halted
+        Halted,
+        /// The minimum voting power threshold is not reached, the block cannot be trusted yet
+        NotEnoughTrust,
+        /// Verification failed, the block is invalid.        
+        InvalidBlock,
+    }
+
+    // NOTE for storage:
+    // - header storage should be a ring buffer (i.e. we keep n last headers, ordered by the insertion time)
+    // - we keep a pointer to the last finalized header (to avoid deserializing the whole buffer)
+    // - insertion moves the pointer and updates the buffer
+    // or?
+    // https://substrate.recipes/ringbuffer.html
+
+    /// Hash of the last imported header from the bridged chain
+    #[pallet::storage]
+    #[pallet::getter(fn get_last_imported_hash)]
+    pub type LastImportedHash<T: Config> = StorageValue<_, BridgedBlockHash, ValueQuery>;
+
+    /// A ring buffer of imported hashes "ordered" by their insertion time
+    #[pallet::storage]
+    pub type ImportedHashes<T: Config> = StorageMap<_, Identity, u32, BridgedBlockHash>;
+
+    /// Current ring buffer position
+    #[pallet::storage]
+    pub(super) type ImportedHashesPointer<T: Config> = StorageValue<_, u32, ValueQuery>;
+
+    /// Bridged chain Headers which have been imported by the client
+    /// Client keeps HeadersToKeep number of these at any time
+    #[pallet::storage]
+    pub(super) type ImportedBlocks<T: Config> =
+        StorageMap<_, Identity, BridgedBlockHash, LightBlockStorage, OptionQuery>;
+
+    // TODO : expose in runtime API and nodes RPC
+    impl<T: Config> Pallet<T> {
+        pub fn get_last_imported_block() -> Option<LightBlockStorage> {
```

Comment:
You could store `LastImportedBlock` instead of `LastImportedHash`, it would simplify the logic in many places and be cheaper in number of storage reads/writes as obv block contains hash. Block takes more storage and is slower to decode but I believe that it's negligible comparing to fetch costs.

Reply:
Actually the headers contain the whole committee, lost of signatures and stuff, so hash is likely to be 100-500 times smaller than a header :/

Reply:
Uhh...so maybe if we implementing those types ourselves, then we could change them and for example don't store everything after the block got validated?

Reply:
I think we still need at least the most recent block with full information :/ (because it seems that the tendermint verifier needs it?). But generally I think one can make multiple optimizations here, it's just at this stage we don't know yet what API will we require from this pallet to build the bridge, so I'm not suggesting any such changes yet...

Reply:
> Actually the headers contain the whole committee, lost of signatures and stuff, so hash is likely to be 100-500 times smaller than a header :/

Is that a problem? We will have to load the last block every time regardless for the purpose of verification. I think that the current implementation has to be more expensive actually since it has to read from multiple storages (pointers, hashes, blocks) and then map to the expected type (from our own `LightBlockStorage` to `tendermint_light_client_verifier::types::LightBlock`).

> I think we still need at least the most recent block with full information :/ (because it seems that the tendermint verifier needs it?)

It seems like we can improve things a bit here. The [`ProdVerifier::verify`](https://docs.rs/tendermint-light-client-verifier/latest/tendermint_light_client_verifier/trait.Verifier.html#tymethod.verify) expects [`TrustedBlockState`](https://docs.rs/tendermint-light-client-verifier/latest/tendermint_light_client_verifier/types/struct.TrustedBlockState.html) which does not contain the signatures - this type is created with [`LightBlock::as_trusted_state(&self)`](https://docs.rs/tendermint-light-client-verifier/latest/src/tendermint_light_client_verifier/types.rs.html#147-154). And we're already doing that. This means, we can have a special storage where we would put the latest trusted state and read that instead.


@fbielejec, can you make the tickets for that work? It doesn't have to hold up this PR.

Reply:
https://cardinal-cryptography.atlassian.net/browse/A0-845

We can't  just store Tendermint type here OC - but we could store something that can be converted into Tendermints TrustedState, sure.

Reply:
> There's one more approach that could decrease the cost of using the pallet: instead of mapping between `*Storage` types and the Tendermint ones we can wrap them with a `struct` that we control:
> 
> ```rust
> struct LightBlockWrapper(tendermint_light_client_verifier::types::LightBlock);
> ```
> 
> and write `Encode`/`Decode` for `LightBlockWrapper`. Now, when we read from storage we no longer have to call expensive `try_into()` as now it's just a wrapper. Additionally, there's no _trying_ anymore since the inner type is exactly what we need.

This won't work IMHO - if the inner type is not Encode/Decode the outer cannot be either. And we cannot impl for Tendermint types becasue Rust, orphaned implementations and all that...

Reply:
>  And we cannot impl for Tendermint external types (not from our crate) because Rust, orphaned implementations and all that...

We can implement that manually though for the outer one, without having separate impls for the inner ones:
```rust
struct LightBlockWrapper(tendermint_light_client_verifier::types::LightBlock);

impl Encode for LightBlockWrapper {
	// Here, we would encode the whole inner `LightBlock` without having separate impls for header, commit, etc.
}
```